<a href="https://colab.research.google.com/github/Sumithrapm/AI-ML/blob/main/Assignment_on_statistics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
from scipy import stats


In [ ]:
# Load dataset
df = pd.read_csv('StudentsPerformance.csv')


In [ ]:
df.head()


,gender,race/ethnicity,parental level of education,lunch,test preparation course,math score,reading score,writing score
0,female,group C,some high school,free/reduced,none,0,17,10
1,female,group B,high school,free/reduced,none,8,24,23
2,female,group B,some high school,free/reduced,none,18,32,28
3,female,group B,some college,standard,none,11,38,32
4,female,group C,some college,free/reduced,none,22,39,33


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


In [ ]:
df.dtypes

,0
gender,object
race/ethnicity,object
parental level of education,object
lunch,object
test preparation course,object
math score,int64
reading score,int64
writing score,int64


In [ ]:
df.isnull().sum()

,0
gender,0
race/ethnicity,0
parental level of education,0
lunch,0
test preparation course,0
math score,0
reading score,0
writing score,0


In [ ]:
df.duplicated().sum()

np.int64(0)

In [ ]:
df.describe()

,math score,reading score,writing score
count,1000.000000,1000.000000,1000.000000
mean,66.001000,69.195000,68.119000
std,15.569567,14.706436,15.226381
min,0.000000,17.000000,10.000000
25%,57.000000,59.000000,58.000000
50%,66.000000,70.000000,69.000000
75%,77.000000,79.250000,79.000000
max,100.000000,100.000000,100.000000


**Calculate the mean, median, and mode for the math, reading, and writing scores**

In [ ]:
print(f"Math score mean= {df['math score'].mean()} ,Reading  Score mean={df['reading score'].mean()}, Writing Score mean={df['writing score'].mean()}",sep="....")
print(f"Math score median= {df['math score'].median()} ,Reading  Score median={df['reading score'].median()} ,Writing Score median={df['writing score'].median()}")
print(f"Math score mode= {df['math score'].mode()[0]} ,Reading  Score mode={df['reading score'].mode()[0]} ,Writing Score mode={df['writing score'].mode()[0]}") # [0] meansfirst in the list else top of list returns

Math score mean= 66.001 ,Reading  Score mean=69.195, Writing Score mean=68.119
Math score median= 66.0 ,Reading  Score median=70.0 ,Writing Score median=69.0
Math score mode= 65 ,Reading  Score mode=72 ,Writing Score mode=74


**2. Determine the range, variance, and standard deviation for each score category**

In [30]:
print(f"Math range= {df['math score'].max()-df['math score'].min()} varience= {df['math score'].var():.2f} std deviation= {df['math score'].std():.2f}")
print(f"Reading range= {df['reading score'].max()-df['reading score'].min()} varience= {df['reading score'].var():.2f} std deviation= {df['reading score'].std():.2f}")
print(f"Writing range= {df['writing score'].max()-df['writing score'].min()} varience= {df['writing score'].var():.2f} std deviation= {df['writing score'].std():.2f}")

Math range= 100 varience= 242.41 std deviation= 15.57
Reading range= 83 varience= 216.28 std deviation= 14.71
Writing range= 90 varience= 231.84 std deviation= 15.23


### **Identify the mean, median, and mode of math scores for students with different parental education levels.**

In [47]:
group_parent = df.groupby('parental level of education')['math score'].agg(['mean', 'median'])

mode_parent = df.groupby('parental level of education')['math score'].agg(lambda x: x.mode()[0]).rename('mode')

# Combine both
final_result = group_parent.join(mode_parent)

print(final_result)

                                  mean  median  mode
parental level of education                         
associate's degree           67.977477    67.0    65
bachelor's degree            69.288136    68.0    63
high school                  61.821429    63.0    57
master's degree              70.254237    73.0    79
some college                 67.128319    68.0    69
some high school             63.134078    65.0    59


## **Compare the mean scores of students who took the test preparation course versus those who didn’t.**

In [52]:
compare_mean=df.groupby('test preparation course')[['math score','reading score','writing score']].mean()
print(compare_mean.round(2))

                         math score  reading score  writing score
test preparation course                                          
completed                     69.97          74.18          74.68
none                          63.79          66.42          64.46


## **Calculate the interquartile range (IQR) for each score category.**

In [60]:
#math IQR,outliers

Q1 = df['math score'].quantile(0.25)
Q3 = df['math score'].quantile(0.75)
IQR = Q3 - Q1
print(f"Q1= {Q1}")
print(f"Q3= {Q3}")
print(f"IQR= {IQR}")


lower = Q1 - 1.5 * IQR
upper = Q3 + 1.5 * IQR

print(f"lower= {lower}")
print(f"upper= {upper}")

outliers = df[(df['math score'] < lower) | (df['math score'] > upper)]
outliers_count = len(outliers)
print(outliers)
outliers_count

Q1= 57.0
Q3= 77.0
IQR= 20.0
lower= 27.0
upper= 107.0
    gender race/ethnicity parental level of education         lunch  \
0   female        group C            some high school  free/reduced   
1   female        group B                 high school  free/reduced   
2   female        group B            some high school  free/reduced   
3   female        group B                some college      standard   
4   female        group C                some college  free/reduced   
5   female        group B                 high school  free/reduced   
6   female        group B            some high school  free/reduced   
7   female        group D          associate's degree  free/reduced   
8   female        group D            some high school  free/reduced   
18  female        group C                 high school  free/reduced   
19  female        group C                some college  free/reduced   
20  female        group C                 high school  free/reduced   
40  female        group 

14

In [67]:
iqr_all = df[['math score','reading score','writing score']].quantile(0.75) - \
                 df[['math score','reading score','writing score']].quantile(0.25)
print(iqr_all)

math score       20.00
reading score    20.25
writing score    21.00
dtype: float64


## **Determine the standard deviation for math scores by lunch type (free/reduced vs. standard).**

In [68]:
std_lunch = df.groupby('lunch')['math score'].std()
print(std_lunch)

lunch
free/reduced    15.882442
standard        13.804560
Name: math score, dtype: float64


## **Find the correlation coefficient between math and reading scores, and then between reading and writing scores.**

In [69]:
corr_math_read = df['math score'].corr(df['reading score'])
corr_read_write = df['reading score'].corr(df['writing score'])

print("Math vs Reading:", corr_math_read)
print("Reading vs Writing:", corr_read_write)

Math vs Reading: 0.8149368617648359
Reading vs Writing: 0.953178417390664



## **Interpret the strength and direction of the correlation. What might this suggest about the relationship between different scores?**

The correlation coefficient tells two things:

1. Direction (positive or negative)
2. Strength (how strong the relationship is)

 From your results:

*   Math & Reading = 0.81
Direction: Positive
Strength: Strong
 This means:

As math scores increase, reading scores also increase
Students who are good in math are generally good in reading

 * Reading & Writing = 0.95
Direction: Positive
Strength: Very Strong

This means:
Reading and writing scores move very closely together
Students strong in reading are almost always strong in writing


### **Calculate the covariance between math and reading scores. How does it compare with the correlation? What does it suggest?**

In [70]:
cov_math_read = df['math score'].cov(df['reading score'])
print(cov_math_read)

186.59840340340335


Covariance → tells direction only

Correlation → tells direction + strength

👉 Covariance (186.60)

Positive value
Shows that math and reading scores increase together

Cov(X,Y)=∑(x−xˉ)(y−yˉ​)/n−1
          
👉 But:

Covariance value is not standardized
It depends on the scale of  raw data (0–100 marks)

👉 Correlation (0.81)

Also positive
Standardized value (between -1 and +1)
Clearly shows strong relationship.

Correlation=​Covariance​/σx​⋅σy

The covariance between math and reading scores is 186.60, which is positive. This indicates that both scores tend to increase together. However, covariance does not provide the strength of the relationship since it is not standardized. The correlation coefficient (0.81) confirms a strong positive relationship. This suggests that students who perform well in math are also likely to perform well in reading


### **Z-Test: Assume the national average math score is 50. Use a one-sample z-test to test if this sample's average math score is significantly different from the national average.**

Sample mean = 66.00 → calculated from df['math score'].mean()
Standard deviation = 15.57 → from df['math score'].std()
Sample size = 1000 → number of rows in your dataset